### ISSUE TITLE
tsc produces invalid typings for parameter destructuring

### ISSUE THREAD

**Issue by uber5001** [Microsoft/TypeScript#3912]
I have an interface that is supposed to have a function signature of:

    (param: {prop: boolean}): any

but I forget the parameter name (I'm not sure if this signature is considered valid):

    ({prop: boolean}): any

When compiling this invalid interface with the `--declaration` flag, an invalid `.d.ts` file results:

    /* foo.ts */
    export interface SomeInterface {
        ({prop: boolean}): any
    }

`tsc foo.ts --module commonjs --declaration`

    /* foo.d.ts */
    export interface SomeInterface {
        ({prop: }: {
            prop: any;
        }): any;
    }

If parameter names are optional, this should produce a valid `.d.ts` file. If parameter names are required, `tsc` should throw an error, or at least warn me that there is an issue.

**Comment by RyanCavanaugh** [comment]
Parameter names are required in some form or another. `{prop: boolean}` is a destructuring parameter (a new ES6 feature) that takes the first argument and assigns its `prop` property into the `boolean` variable.

See #2023 for the bug about destructured parameters creating invalid .d.ts files

**Comment by uber5001** [comment]
Hey Ryan. I don't think this is a duplicate of #2023, as I'm not attempting to destructure anything here. Since parameter names are required (thank you for clarifying!), the issue is now that `tsc` does not throw any errors, and silently produces invalid output.

This was tested using the latest version of tsc, taken from master/bin/tsc. Some sort of error should be produced at compile-time instead of generating invalid output, right?

**Comment by RyanCavanaugh** [comment]
You're right. Thanks for following up -- I misread the linked bug.

**Comment by ahejlsberg** [comment]
Sadly, this is not a bug. The ES6 destructuring syntax is eerily similar to the syntax for object types. The following is perfectly valid ES6:

```typescript
function foo({ prop: boolean }) {
    let x = boolean;  // Destructuring declares local named boolean of type any
}

foo({ prop: 42 });
```

Not entirely clear how we'd make something like that an error.

**Comment by mhegazy** [comment]
Just to clarify, the issue here is that the declaration file generated is not valid. there is an extra ":" emitted after "prop" in:

```ts
interface SomeInterface {
    ({prop: }: {
        prop: any;
    }): any;
}
```
